OpenSim/Common/TableSource.h

```diff
@@ -95,21 +121,52 @@ class TableSource_ : public Component {
         return _table;
     }
 
-    /** Replace the existing TimeSeriesTable_ this TableSource_ currently 
-    holds.                                                                    
+    /** Replace the existing TimeSeriesTable_ that this TableSource_ currently 
+    holds. This operation is not allowed if TableSource_ already contains a
+    value for the property `filename`. In such cases, update the property
+    'filename' using `upd_filename()` to empty string and call this function.
 
-    \throws KeyNotFound If table provided does not have column-labels.        */
+    \throws InvalidCall If property `filename` is set. This call is not allowed
+                        if `filename` property is set.                        
+    \throws KeyNotFound If table provided does not have column labels.        */
     void setTable(const Table& table) {
+        OPENSIM_THROW_IF(!get_filename().empty(),
+                         InvalidCall,
+                         "Property 'filename' is set. Cannot set Table now.");
+
+        setTable_impl(table);
+    }
+
+    /// @}
+
+protected:
+    /** Replace the existing TimeSeriesTable_ this TableSource_ currently holds
+    with the table `tablename` read from the file `filename`. `filename` and 
+    `tablename` are properties of TableSource_. In order to update the table
+    from a file, update the properties using `upd_filename()` and 
+    `upd_tablename()` and call this function.
+
+    Exceptions thrown include the ones thrown by TimeSeriesTable_ constructor
+    that takes a filename and tablename.
+
+    \throws InvalidCall If the property 'filename' is empty.
+    \throws KeyNotFound If table provided does not have column-labels.        */
+    void setTable() {
+        OPENSIM_THROW_IF(get_filename().empty(),
+                         InvalidCall,
+                         "Property 'filename' is empty.");
+
+        setTable_impl(TimeSeriesTable_<ET>{get_filename(), get_tablename()});
+    }
```

Comment:
The is only valid/necessary inside `extendFinalizeFromProperties`, so why not make this call directly there? What is the extra indirection buying here?